gql_http_link/lib/gql_http_link.dart

```diff
@@ -20,94 +55,101 @@ import "package:meta/meta.dart";
     _httpClient = httpClient ?? http.Client();
   }
 
-  /// Serializes the request
-  ///
-  /// Extend this to add non-standard behavior
-  @visibleForOverriding
-  Map<String, dynamic> serializeRequest(Request request) => <String, dynamic>{
-        "operationName": request.operation.operationName,
-        "variables": request.operation.variables,
-        "query": printNode(request.operation.document),
-      };
-
-  /// Parses the response body
-  ///
-  /// Extend this to add non-standard behavior
-  @visibleForOverriding
-  Response parseResponse(Map<String, dynamic> body) => Response(
-        errors: (body["errors"] as List)
-            ?.map(
-              (dynamic error) => parseError(error as Map<String, dynamic>),
-            )
-            ?.toList(),
-        data: body["data"] as Map<String, dynamic>,
-      );
+  @override
+  Stream<Response> request(
+    Request request, [
+    NextLink forward,
+  ]) async* {
+    dynamic body;
 
-  /// Parses a response error
-  ///
-  /// Extend this to add non-standard behavior
-  @visibleForOverriding
-  GraphQLError parseError(Map<String, dynamic> error) => GraphQLError(
-        message: error["message"] as String,
-        path: error["path"] as List,
-        locations: (error["locations"] as List)
-            ?.map(
-              (dynamic error) => parseLocation(error as Map<String, dynamic>),
-            )
-            ?.toList(),
-        extensions: error["extensions"] as Map<String, dynamic>,
+    try {
+      body = json.encode(
+        serializer.serializeRequest(request),
       );
-
-  /// Parses a response error location
-  ///
-  /// Extend this to add non-standard behavior
-  @visibleForOverriding
-  ErrorLocation parseLocation(Map<String, dynamic> location) => ErrorLocation(
-        line: location["line"] as int,
-        column: location["column"] as int,
+    } on Exception catch (e) {
+      throw SerializerException(
+        originalException: e,
       );
+    }
 
-  Future<Response> _post(Request request) async {
-    final response = await _httpClient.post(
-      graphQLEndpoint,
+    final httpResponse = await _httpClient.post(
+      uri,
       headers: {
         "Content-type": "application/json",
-        "Accept": "application/json",
+        "Accept": "*/*",
+        ...defaultHeaders,
+        ..._getHttpLinkHeaders(request),
       },
-      body: json.encode(
-        serializeRequest(request),
-      ),
+      body: body,
     );
 
-    final dynamic body = json.decode(
-      utf8.decode(
-        response.bodyBytes,
+    Response response;
+
+    try {
+      final dynamic responseBody = json.decode(
+        utf8.decode(
+          httpResponse.bodyBytes,
+        ),
+      );
+
+      response = parser.parseResponse(responseBody as Map<String, dynamic>);
+    } on Exception catch (e) {
+      throw ParserException(
+        originalException: e,
+        response: httpResponse,
+      );
+    }
+
+    if (httpResponse.statusCode >= 300 ||
```

Comment:
Status codes >= 400 < 500 are technically client errors. Do you think a new Exception should be thrown to make it easier to handle? (Or renaming ServerException to HTTPException or something like this)

Reply:
This time I'm going with an equal behavior to `apollo-link-http` regarding the status codes.

I'm going with `ServerException` because I'm thinking it (and other exceptions) would eventually used across many kinds of links. For example, Link which resolves queries on the client-side could use `ServerException` to say that there was an exception in the GraphQL server (located in the client) when some exception happens during query execution.